kotlinx-coroutines-core/jvm/test/linearizability/SegmentQueueSynchronizerLCStressTests.kt

```diff
@@ -0,0 +1,770 @@
+    private var count = initialCount
+    private val waiters = ArrayList<CancellableContinuation<Unit>>()
+
+    fun countDown() {
+        if (--count == 0) {
+            waiters.forEach { it.tryResume0(Unit) }
+            waiters.clear()
+        }
+    }
+
+    suspend fun await() {
+        if (count <= 0) return
+        suspendAtomicCancellableCoroutine<Unit> { cont ->
+            waiters.add(cont)
+        }
+    }
+
+    fun remaining(): Int = count.coerceAtLeast(0)
+
+    override fun extractState() = remaining()
+}
+
+
+// ###########################
+// # BARRIER SYNCHRONIZATION #
+// ###########################
+
+/**
+ * This synchronization primitive allows a set of coroutines to
+ * all wait for each other to reach a common barrier point.
+ *
+ * The implementation is straightforward: it maintains a counter
+ * of arrived coroutines and increments it in the beginning of
+ * [arrived] operation. The last coroutines should resume all the
+ * previous ones.
+ *
+ * In case of cancellation, the handler decrements the counter if
+ * not all the parties are arrived. However, it is impossible to
+ * make cancellation atomic (e.g., Java's implementation simply
+ * does not work in case of thread interruption) since there is
+ * no way to resume a set of coroutines atomically. Thus,
+ * this implementation is non-atomic if cancellation happens
+ * simultaneously to the last [arrive], but is correct under
+ * the strong cancellation model, when continuation can be
+ * cancelled if it is logically resumed but not dispatched yet.
+ */
+internal class Barrier(private val parties: Int) : SegmentQueueSynchronizer<Unit>() {
+    override val resumeMode get() = ASYNC
+    override val cancellationMode get() = SMART_ASYNC
+
+    // The number of coroutines arrived to this barrier point.
+    private val arrived = atomic(0L)
+
+    /**
+     * Waits for other parties and returns `true`.
+     * Fails if this invocation exceeds the number
+     * of parties, returns `false` in this case.
+     */
+    suspend fun arrive(): Boolean {
+        // Increment the number of arrived parties.
+        val a = arrived.incrementAndGet()
+        return when {
+            // Should we suspend?
+            a < parties -> {
+                suspendCoroutine<Unit> { cont -> suspend(cont) }
+                true
+            }
+            // Are we the last party?
+            a == parties.toLong() -> {
+                // Resume all waiters.
+                repeat(parties - 1) {
+                    resume(Unit)
+                }
+                true
+            }
+            // Should we fail?
+            else -> false
+        }
+    }
+
+    override fun onCancellation(): Boolean {
+        // Decrement the number of arrived parties if possible.
+        arrived.loop { cur ->
+            // Are we going to be resumed?
+            // The resumption permit should be refused in this case.
+            if (cur == parties.toLong()) return false
```

Comment:
A small problem here. I believe this should be `cur <= parties.toLong()`, but I think it's interesting why so and, more importantly, why it didn't cause any problems.

`parties - 1` calls happen and go to sleep; then comes the call that is about to wake everyone up but doesn't yet do anything after incrementing the counter; then, 1000 new calls happen, and `a == parties + 1000`. Then, every thread is cancelled, performs `onCancellation` and, since `arrived` is much bigger than `parties.toLong()`, successfully decrements `arrived`. Then the resumer comes, sees `parties - 1` `CANCELLED` marks and proceeds to wake up `parties - 1` more cells.

This works only because this behavior, while not optimal, doesn't break anything: after the cells are all woken up, this SQS is never used again. However, if `resumeMode`  was set to `SYNC`, everything would hang in this scenario, but it seems that with the change above it would work fine in this case too.

It should be noted though that without the check for `cur == parties.toLong()` the code would not be linearizable: as is, it guarantees that once the barrier is broken, it can't be accidentally "un-broken" no matter how many cancellations happen, and so the subsequent calls to `arrive` all return `false` in any case.

Reply:
It should be either `>=` or `==`. In my opinion, it is always better to decrement the counter if possible, so that the number of `resume` invocations is decreased. We should also add a check that the barrier is completed in the beginning.